### PR TITLE
Add `fn.experimental.audio_resample` GPU

### PR DESCRIPTION
**Pull request by jantonguirao**
## Category:

 **New feature**

## Description:

- It adds a new operator `fn.experimental.audio_resample` for the GPU backend.
- The operator builds on top of the existing CPU operator counterpart
- The implementation relies on the GPU signal resampling kernel.

### Affected modules and functionalities:

New Op

### Key points relevant for the review:

Test correctness?

### Tests
- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [x] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: ARES.[01-08]

**JIRA TASK**: DALI-1445

### CODE REVIEW

File: dali/test/python/test_operator_audio_resample.py [hunk 1]
```diff
@@ -34,64 +34,76 @@
 rates = [ 16000, 22050, 12347 ]
 lengths = [ 10000, 54321, 12345 ]
 
-def create_test_files():
```
Comment (jantonguirao) [inline dali/test/python/test_operator_audio_resample.py:37]: Note: nose was picking this as a test

Comment (mzient) [reply dali/test/python/test_operator_audio_resample.py:37]: Making it "private" (with leading underscore) would help, too.

File: dali/operators/audio/resample_gpu.cc [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        ResampleTyped<Out, In>(view<Out>(out), view<const In>(in), ws.stream());
+      ), (  // NOLINT
+        DALI_FAIL(
+          make_string("Unsupported input type: ", in.type(), "\nSupported types are : ",
+                      ListTypeNames<AUDIO_RESAMPLE_TYPES>()));
+      ));  // NOLINT
+    ), (  // NOLINT
+      DALI_FAIL(
+        make_string("Unsupported output type: ", dtype_, "\nSupported types are : ",
+                    ListTypeNames<AUDIO_RESAMPLE_TYPES>()));
```
Comment (mzient) [inline dali/operators/audio/resample_gpu.cc:54]: Can you make this check common to both backends, somehow, perhaps in Setup? Then here you could just do `assert(!"Unreachable code");`. Now we have the messages with supported types in four places (input, output, two backends).

File: dali/operators/audio/resample.h [hunk 3]
```diff
@@ -144,7 +148,8 @@ class ResampleBase : public Operator<Backend> {
   ArgValue<float> scale_{"scale", spec_};
   ArgValue<int64_t> out_length_{"out_length", spec_};
 
-  std::vector<double> scales_;
+  using Args = kernels::signal::resampling::Args;
+  SmallVector<Args, 128> args_;
```
Comment (mzient) [inline dali/operators/audio/resample.h:152]: Plain vector will do just fine. Large SmallVectors (however weird that sounds) are better suited for temporary local buffers, where the difference between (frequent) stack allocation (free) and heap allocation (thousands of cycles) is of essence. Here, the vector will be reallocated a few times per operator lifetime at worst (typically it will be allocated just once) - and chances are, we'll still run over 128.
```suggestion
  std::vector<Args> args_;
```

File: dali/kernels/signal/resampling_gpu.cu [hunk 4]
```diff
@@ -77,13 +77,12 @@ void ResamplerGPU<Out, In>::Run(KernelContext &context, const OutListGPU<Out> &o
     desc.out = out_sample.data;
     desc.window = window_gpu_;
     const auto &in_sh = in_sample.shape;
-    const auto &out_sh = out_sample.shape;
```
Comment (jantonguirao) [inline dali/kernels/signal/resampling_gpu.cu:80]: removed this because it was seen as not-used in release builds (therefore a warning)

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+# Copyright (c) 2020-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (mzient) [inline dali/test/python/test_torch_pipeline_rnnt.py:1]: Shouldn't this be a separate PR?

File: dali/test/python/test_operator_audio_resample.py [hunk 6]
```diff
@@ -34,64 +34,73 @@
...
+      if not np.allclose(out_arr, ref, 1e-6, eps):
+        print("Actual: ", out_arr)
+        print(out_arr.dtype, out_arr.shape)
         print("Reference: ", ref)
         print(ref.dtype, ref.shape)
-        print("Diff: ", out.at(i).astype(np.float) - ref)
-        assert np.allclose(out.at(i), ref, 1e-6, eps)
-
+        print("Diff: ", out_arr.astype(np.float) - ref)
+        assert False
```
Comment (mzient) [inline dali/test/python/test_operator_audio_resample.py:103]: Nitpick: the original code deliberately repeated the check here, so that the error in nosetests would appear as more than `False` in non-verbose runs.

File: dali/operators/audio/resample.h [hunk 7]
```diff
@@ -125,7 +128,8 @@ class ResampleBase : public Operator<Backend> {
           DALI_FAIL(make_string("Cannot produce a non-empty signal from an empty input.\n"
             "Error at sample ", s));
         }
-        scales_[s] = in_length ? 1.0 * out_length / in_length : 0.0;
+        args_[s].in_rate = 1.0;
+        args_[s].out_rate = in_length ? 1.0 * out_length / in_length : 0.0;
```
Comment (mzient) [inline dali/operators/audio/resample.h:132]: Now that we have `args`, this would increase precision:
```suggestion
        args_[s].in_rate = in_length ? in_length : 1;    // avoid division by 0
        args_[s].out_rate = out_length ? out_length : 1;  // avoid division by 0
```

**Review by mzient — APPROVED** [review]

File: dali/operators/audio/resample.h [hunk 8]
```diff
@@ -125,7 +128,8 @@ class ResampleBase : public Operator<Backend> {
           DALI_FAIL(make_string("Cannot produce a non-empty signal from an empty input.\n"
             "Error at sample ", s));
         }
-        scales_[s] = in_length ? 1.0 * out_length / in_length : 0.0;
+        args_[s].in_rate  = in_length  ? in_length  : 1;  // avoid division by 0
+        args_[s].out_rate = out_length ? out_length : 1; // avoid division by 0
```
Comment (mzient) [inline dali/operators/audio/resample.h:132]:
```suggestion
        args_[s].out_rate = out_length ? out_length : 1;  // avoid division by 0
```
the linter will complain

File: dali/test/python/test_operator_audio_resample.py [hunk 9]
```diff
@@ -34,64 +34,73 @@
 rates = [ 16000, 22050, 12347 ]
 lengths = [ 10000, 54321, 12345 ]
 
-def create_test_files():
+def create_files():
```
Comment (awolant) [inline dali/test/python/test_operator_audio_resample.py:37]: I think this would be another way, more verbose probably:

```suggestion
from nose.tools import nottest
@nottest
def create_files():
```

**Review by awolant — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build
